lookout/style/format/rules.py

```diff
@@ -31,17 +31,116 @@
+    """
+    A modelforge model to store Rules instances.
+    It is required to store all the Rules for different programming languages in a single model,
+    named after each language.
+    Note that Rules must be fitted and Rules.base_model is not saved.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._rules_by_lang = {}
+
+    @property
+    def languages(self):
+        return sorted(self._rules_by_lang.keys())
+
+    def construct(self, ruless: Iterable[Tuple[str, "Rules"]]) -> "FormatModel":
+        for name, rules in ruless:
+            self[name] = rules
+        return self
 
     def dump(self) -> str:
-        pass
+        if len(self) == 0:
+            return "<empty FormatModel>"
+        language = self.languages[0]
+        param_names = self[language]._get_param_names()
+        return "Models number: %d.\n" \
+               "First model name: %s\n" \
+               "First model params: %s\n" \
+               "Rules number: %d.\n" % \
+               (len(self),
+                language,
+                ", ".join(["%s=%s" % (p, str(getattr(self[language], p))) for p in param_names]),
+                len(self[language]._rules))
 
     def _generate_tree(self) -> dict:
-        pass
+        languages = self.languages
+        return dict(
+            languages=languages,
+            paramss=[self[lang].get_params(deep=False) for lang in languages],
+            ruless=[self._disassemble_rules(self[lang]._rules) for lang in languages],
+        )
 
     def _load_tree(self, tree: dict) -> None:
-        pass
+        for name, params, rules in zip(tree["languages"], tree["paramss"], tree["ruless"]):
+            params = dict(params)
+            params["top_down_greedy_budget"] = tuple(params["top_down_greedy_budget"])
+            params["_rules"] = self._assemble_rules(rules)
+            _rules = Rules.__new__(Rules)
+            _rules.__setstate__(params)
+            self[name] = _rules
+
+    def __len__(self) -> int:
+        return len(self._rules_by_lang)
+
+    def __getitem__(self, lang: str) -> "Rules":
+        """
+        Get a Rules estimator by its language.
+        :param lang: Estimator language.
+        :return: Rules estimator instance.
+        """
+        return self._rules_by_lang[lang]
+
+    def __setitem__(self, lang: str, rules: "Rules"):
+        """
+        Set a new Rules estimator to the model by its language.
+        """
+        if not rules.fitted:
+            raise NotFittedError("Rules estimator should be fitted before adding to FormatModel.")
+        self._rules_by_lang[lang] = rules
+
+    def __iter__(self):
+        yield from self._rules_by_lang.__iter__()
+
+    def __contains__(self, item):
+        return item in self._rules_by_lang
+
+    @staticmethod
+    def _assemble_rules(rules_tree: dict) -> List[Rule]:
+        rules = []
+        cur_length = 0
+        rule_attrs = [RuleAttribute(*params) for params in
```

Comment:
You can change to `()` instead of `[]` - generator. Then inside the loop:
```python
rules.append(Rule(tuple(islice(rule_attrs, length)), RuleStats(cls, conf)))
```
Thus you can drop `cur_length`.